### ISSUE TITLE
reactions should not be used to mark posts as spam/unwanted

### ISSUE THREAD

**Issue by eskema** [nostrability/nostrability#88]
## client initiating action
not sure but I think Amethyst, maybe others as well

## client receiving or observing action (or lack thereof)
any

## what happens 
clients using reactions to mark posts using the emoji ⚠️
this behaviour does the exact oposite as it spreads the unwanted post to clients that don't use this dumb and obscure technique.

## suggestion
don't use reactions of kind:7, use a another kind so there is no ambiguous meaning

**Comment by staab** [comment]
Yes, this is super bad, and has been used in the past to flag illegal content, publicly and irrevocably association people's accounts with CSAM. Whoever is doing this should stop. Use 1984 at least, but even those are public. I currently use nos' taggr bot in Coracle for reports.

**Comment by jb55** [comment]
people might be tempted to use `-` for this as well. plz no.

**Comment by alltheseas** [comment]
@vitorpamplona 👀 see @eskema report. Is it amethyst?

==

Tangent:

I'm tinkering with nostur @fabianfabian - I see that in certain notes there is grey context that indicates a note was posted e.g. "via nostrudel". 

cc @hzrd149 

Is this is an opt-in spec/NIP/best practice? Having posted via client XYZ context makes nostrability job a bit easier.

**Comment by vitorpamplona** [comment]
> this behaviour does the exact oposite as it spreads the unwanted post to clients that don't use this dumb and obscure technique.

Why does it spread? 

Yes, Amethyst does that. It's a nice way to let everyone know that the user has reported the content. 

The user will send a negative reaction anyway. We just do the two at the same time for them.

**Comment by alltheseas** [comment]
1984 reports have a singular, defined use case. Some apps ingest these, and e.g. warn their users of reports. 

Does Amethyst or any other app ingest ⚠️ warning emojis, and warn their users of notes with this?

**Comment by vitorpamplona** [comment]
Nothing custom for ⚠️. It's just a reaction. The interpretation is up for grabs like with any other reaction.

Like ⚠️, there are hundreds of other negative reactions that people are using to send each other signals. In fact, I commonly see reactions that are positive in some countries and negative in others. So, it's all about the user judgement when they see the reaction set for a post.

**Comment by alltheseas** [comment]
> it's all about the user judgement when they see the reaction set for a post.

Here is the user flow I envision for 1984, and ⚠️:

1984 user flow

1. User A reports 1984 
2. User B app receives 1984 report
3. User B app does nothing, or warns user and e.g. blurs post
4. Assuming User B app warns user B, user B decides to proceed to read the flagged note anyway, or move on

Warning sign user flow ⚠️

1. User A reacts with ⚠️ to noteX (assume user B app does not receive 1984 report for noteX)
2. User B app receives ⚠️ react and adds to noteX reaction tally
3. User B finds noteX in timeline
4. User B reads note
5. User B sees reaction count
6. User B sees ⚠️ as one of Y reactions in detailed reaction view

In the warning sign user flow, I am not certain about the order of steps 4. - 6. Say the order is as described. That is, in the case user B reads the note before going to detailed reaction view, will have the warning reaction served its intended cautionary purpose?

**Comment by vitorpamplona** [comment]
The goal for ⚠️ is not to anticipate the read or mark as something less. It is to participate in the rally of reactions together with other reactions. 

It's not any different from the other negative reactions people are using, like 😡🤬😰😨😮😖😣😵🥵🤢🤮👎

Whoever gets to take a look at the post should be able to quickly analyze how well it was received by just looking at them. The point of ⚠️ is just to add the appropriate participation in place.

**Comment by staab** [comment]
@vitorpamplona I think you should reconsider this. Obe of your users contacted me using a burner key because he had taken it upon himself to "report" CSAM via Amethyst. His "reports" then showed up under his "likes" tab on coracle. This is of course a problem with kind 7 being overloaded in the first place between binary reactions and emoji reactions, but further overloading it with reports makes the problem much worse.

**Comment by vicariousdrama** [comment]
> clients using reactions to mark posts using the emoji ⚠️ this behaviour does the exact oposite as it spreads the unwanted post to clients that don't use this dumb and obscure technique.

BoostZapper reacts with the ⚠️ emoji as diagnosis when it is unable to zap a user for their post, often due to a user configuration error regarding their LUD06 or LUD16, but can also occur if the value set is a blocked account. It's a positive indication that the service saw the user's reply, and no further action will be taken.

**Comment by eskema** [comment]
> The goal for ⚠️ is not to anticipate the read or mark as something less. It is to participate in the tally of reactions together with other reactions.
> 
> It's not any different from the other negative reactions people are using, like 😡🤬😰😨😮😖😣😵🥵🤢🤮👎
> 
> Whoever gets to take a look at the post should be able to quickly analyze how well it was received by just looking at them. The point of ⚠️ is just to add the appropriate participation in place.

Reactions are participative like you say, but I don't think in this case it is. I'm not sure how it happens on Amethyst UI, is it the same as any other emoji? or is there a button for report / flag as spam? is the user actively choosing that specific emoji from all emojis or is amethyst applying it after the user clicks a specific button? 

when a client makes a request for notes their follows reacted to, they generally don't expect to be sent posts flagged as spam, those should be on their own kind (1984 or something else). I don't think any user using this really understands what's happening under the hood and that the posts they flag are showing up on other clients BECAUSE of them.

You are under the assumption that your way of displaying content is universal when it most definitely isn't. 

Not all clients are twitter/social/mobile with the same UI, my client doesn't tally likes, they display them as any other event.
there should not be any expectation of further flow happening when you send a reaction event. it should just be that, a simple reaction that you expect the world to see, if on the other hand you expect the content to be blocked and hidden from view, then a reaction is clearly the opposite

**Comment by vitorpamplona** [comment]
> @vitorpamplona I think you should reconsider this. One of your users contacted me a while back using a burner key because he had taken it upon himself to "report" CSAM via Amethyst. His "reports" then showed up under his "likes" tab on coracle. This is of course a problem with kind 7 being overloaded in the first place between binary reactions and emoji reactions, but further overloading it with reports makes the problem much worse.

That's a problem of making a "likes" tab that doesn't filter negative reactions, not a problem of using ⚠️ this way. 

Don't assume reactions are positive. They are not. They never were and never will be. 

Assuming every reaction is a like is wrong and listing them on a profile indiscriminately doesn't work.

**Comment by eskema** [comment]
> > @vitorpamplona I think you should reconsider this. One of your users contacted me a while back using a burner key because he had taken it upon himself to "report" CSAM via Amethyst. His "reports" then showed up under his "likes" tab on coracle. This is of course a problem with kind 7 being overloaded in the first place between binary reactions and emoji reactions, but further overloading it with reports makes the problem much worse.
> 
> That's a problem of making a "likes" tab that doesn't filter negative reactions, not a problem of using ⚠️ this way.
> 
> Don't assume reactions are positive. They are not. They never were and never will be.
> 
> Assuming every reaction is a like is wrong and listing them on a profile indiscriminately doesn't work.

I'm not assuming that every reaction is positive, but that the reacted content will be spread out along with your reaction to it. when you react, you want to say to the world what you think of that content, it's very different than wanting to hide that content.

**Comment by vitorpamplona** [comment]
We only mark ⚠️ when it's a report (the user wants to tell the world), not when it's a simple hide.

**Comment by eskema** [comment]
<img width="1496" alt="Screenshot 2024-09-08 at 12 33 35" src="https://github.com/user-attachments/assets/9dc95d58-0e47-4ae9-80ca-f8a6c9e98626">

do you think the user here wants the world to know this is spam or is he trying to hide it from other users?

**Comment by vitorpamplona** [comment]
I don't know what this is. But a user reacting to ⚠️ or 🤮 is telling the world he/she didn't like the post (with slightly different meanings) and that's how you should interpret it.

**Comment by eskema** [comment]
yes, but it's not what is happening here, the user is clicking a report button, and the app is sending a reaction note instead of a proper report, so the user has no idea of what happened because your app does it in the background with a different flow than the user choosing 🤮 to actually react to content and express his public opinion.

**Comment by vitorpamplona** [comment]
> the app is sending a reaction note instead of a proper report

The app always sends both when reporting. It's not one instead of another. The report helps the NIP-56 systems out there filter content out. The negative reaction contributes to the semantical analysis of the reaction set the post gets.

**Comment by staab** [comment]
> That's a problem of making a "likes" tab that doesn't filter negative reactions, not a problem of using ⚠️ this way.

Yes, but also yes it is. First, it's impossible to classify every emoji as positive or negative. But that's not the problem at issue here. It's that given the semantics of reactions, it's reasonable to show them, and the content they're attached to. If the content is to be suppressed, that is the opposite intention.

Also, I think you're looking at the communication that happens to emojis in too simplistic a way. A 🤮 emoji could indicate positive sentiment, for example sympathy with the note's author. Likewise, ⚠️ could indicate agreement with the note, like if the author was pointing out suspicious behavior.

I mean, the big point here is that people are being so unpleasantly surprised by what you're doing with kind 7s that they are deleting their accounts. Kind 7 should at the very least be non-toxic.

**Comment by staab** [comment]
Another thing to consider is that you're also mixing user provided content with application semantics, which is just a generally bad idea in software. Imagine if JsSON parsers removed any object containing "hidden: true". You would no longer be able to encode that data. What you're doing makes it impossible to encode a particular emoji without dragging along application semantics.

**Comment by vitorpamplona** [comment]
I think we can agree to disagree. 

> It's that given the semantics of reactions, it's reasonable to show them

Sure. Everything can be reasonable to show. You can show the user's reports as well. Showing reactions just presents everything the user has liked and disliked in one big list. 

The receiving user can hide reacted posts that have also been reported to. That's the decision of the receiving user, not the sender's. 

> If the content is to be suppressed, that is the opposite intention.

This idea that a reaction makes a given content "more accessible" just because the client doesn't process what the reaction actually is is ridiculous to me and the wrong implementation of the spec. If you can't process if a reaction is positive or negative, you shouldn't give more visibility to the post just based on the fact people is reacting to it. 

Also, as a user, reposts give more visibility to posts. Not blind reactions. I think you are mixing their semantics. 

> I mean, the big point here is that people are being so unpleasantly surprised by what you're doing with kind 7s that they are deleting their accounts. Kind 7 should at the very least be non-toxic.

That feels like a rare overreaction and a misunderstanding of user expectations. Those expectations should be fixed. 

> Another thing to consider is that you're also mixing user provided content with application semantics

You are wrong. There is no additional semantics that we build around the warning emoji. Just because the app does something automatically for our users, it doesn't mean that this is additional semantics. It's just a quick action helper that does the two things at the same time. The meaning behind it is entirely up to the receiver and warning signs from Amethyst should be interpreted as if a human reacted with the warning emoji, whatever that may mean to the receiving client.

**Comment by staab** [comment]
> If you can't process if a reaction is positive or negative, you shouldn't give more visibility to the post just based on the fact people is reacting to it.

As I mentioned, it's impossible to reduce the srmantics of a high-definition subjective medium to a low-definition objective one. As for visibility, this is how people use reactions - they are used as signal for trending feeds in multiple places.

> The receiving user can hide reacted posts that have also been reported to. That's the decision of the receiving user, not the sender's.

You invented these semantics, and now you're asking everyone else to sownload more data to protect their users from objectionable content?

> That feels like a rare overreaction and a misunderstanding of user expectations. Those expectations should be fixed.

Then you won't mind surveying your users to get their reaction?

> There is no additional semantics that we build around the warning emoji. Just because the app does something automatically for our users, it doesn't mean that this is additional application semantics.

Then I don't see the point. If there is already a report being sent, what does the emoji add?

**Comment by alltheseas** [comment]
As Nos team takes moderation as a high priority, what do you think @setch-l?

**Comment by fabianfabian** [comment]
> this behaviour does the exact oposite 

I can confirm at least for Nostur that it does indeed do the opposite and could make those posts appear in the Hot/Discover feed so I would have to make a filter for positive vs negative emojis if this becomes a problem.

> I'm tinkering with nostur @fabianfabian - I see that in certain notes there is grey context that indicates a note was posted e.g. "via nostrudel".

This is NIP-89 client tag: https://github.com/nostr-protocol/nips/blob/master/89.md#client-tag

**Comment by staab** [comment]
@vitorpamplona just to be clear, clients that show likes rather than emojis have a problem (coracle included), because the consensus on kind 7 that has emerged is that they are "emoji reactions". It is impossible now to reduce kind 7 to binary likes, or to introduce a new spec that is only for binary likes, because you would lose all signal coming from reaction-based clients, which would be a meaningful fork. Us like-based clients will have to deal with that somehow, but the issue here is react/report conflation.

**Comment by vitorpamplona** [comment]
> consensus on kind 7 that has emerged is that they are "emoji reactions"

Emoji reactions have been there since @jb55 introduced the NIP on Aug 22. It didn't "emerge" over time. In fact, I remember that before Amethyst all clients that I was using at the time had emoji reactions and never actually used the word "like" because it was such a hated word back then. 

> the issue here is react/report conflation

Maybe it helps if I tell the story of why Amethyst does that. The warning reaction sign was added 3 months before reports existed, in Feb 2022. It was the main way users were telling each other that some content was bad. In fact, many users had their own system of negative reactions that used separate emojis for different types of reports. When NIP-56 was introduced, which was derived by a requirement from the Apple store, Amethyst added it, but users kept reacting as well. So, to help them out, we made one interface that did both. Over time the reports started to be used by relay operators to filter content down while negative reactions were used by content indexers to slightly increase and decrease the score of each post. 

So, in my mind, there is no conflation what so ever. The two things can and should happen in parallel because they are feeding different systems.

**Comment by alltheseas** [comment]
It would be useful for us to gather more input from the nostrich end users of Amethyst, and other apps. I suspect many will not care to chime in, and the valuable information will be from nostrich end users who care enough to comment.

@eskema made a post on this - I encourage yall to quote, add to, and/or make your own post on this question

https://njump.me/nevent1qvzqqqqqqypzp8k85aupv7hmr5cvfqeaayezmgxq3wn3560pjyw427xnz39m2ephqqs2shexcnpmqa5w8v5cal78sad0js3em5ezx62yhjvet863fmpnjcgg8qh4s

**Comment by staab** [comment]
> It didn't "emerge" over time.

Right, the original spec supported both +/- and emojis. What I mean is that emojis have won. Binary likes can be represented as emojis, but emojis can't be represented as binary likes. The overload on the original spec was a mistake. But that's my problem, as a "like" client, and still beside the point I think.

> Maybe it helps if I tell the story of why Amethyst does that.

Thanks for the reminder, I had forgotten the history. However, you're contradicting yourself:

> The two things can and should happen in parallel because they are feeding different systems with separate semantics.

> There is no additional semantics that we build around the warning emoji.

Which is it? Regardless of the history, I would be very surprised if there are still people relying on the warning emoji as a signal independent of NIPs 36 and 56. I think all my points still apply — there is real damage being done by the warning emoji, because people do view feeds of kind 7s in several different clients, in several different ways. It's also unavoidably an overloading of user-generated content with application semantics.

**Comment by staab** [comment]
I've published my own survey: `nostr:nevent1qyw8wumn8ghj76r0v3kxymmy9e3k7unpvdkx2tn5dahkcue0qy2hwumn8ghj7un9d3shjtnyv9kh2uewd9hj7qg6waehxw309ac8junpd45kgtnxd9shg6npvchxxmmd9uq3gamnwvaz7tm0dek8jmn0w3jhxtnvdakz7qpqsvzt0mjhhkymjexca20xvrpsr36mkgpt0ttph8hycnwvarnxnj0qdvnk0g`

**Comment by setch-l** [comment]
The outcome we want is the ability for our apps to support user-led moderation. We want this because the centralized moderation systems today make assumptions about what we all want to see or don't want to see and/or are driven by advertisers / investor desires. 

If we build workflows that assume what our users intend or don't intend then we are not much better than the platform social media apps. We're just a different flavor of centralization. From our experience at Nos, I would NOT recommend using emoji's to indicate spam or any NIP56 categories. Tagging content or accounts has real reputation consequences and should be deliberately done and not assumed at any part of the process. 

What we've done in Nos to resolve some of these issues is to ask the user to confirm at each step of the way what they want to do: 
Step 1: What NIP56 category do you want use to tag the content? or which NIP56 category do you want to use to tag the user/account? 
Step 2: Do you want to post the tag publicly or do you want Nos to tag it for you? (This is to prevent dogpiling by the account that was tagged)
Step 3: The note that was tagged in the app then has an overlay indicating why it was tagged w/ the option for other users to view it. 

The reason we add all of these steps is to prevent abuse of the user-led moderation by bad actors. Bad actors are usually lazy, they do not want to manually mess with people they want to automate it. The concern i would have with the error / spam emoji is that it could be easily used for abuse by a bad actor. 

I saw that another app in this thread is using the error emoji to indicate that a zap transaction did not go through. I would also recommend against this usage and instead tell the user - hey this didn't work. By being clear, we can help improve the overall Lightning network of transactions and potentially prevent a lot of back and forth of I sent it to you, no you didn't etc. I would say a failed zap could be a notification possibly - but I would research that with folks who send a lot of zaps... 

From user research I've seen an interest in emoji's beyond likes in part to get away from the dopamine machine experience. 

In conclusion: 
Yes to all sorts of emoji reactions! 
No to emoji's doing any sort of content labeling.
No to emoji's indicating failed actions.

**Comment by alltheseas** [comment]
Learning from a user that they are surprised about and have been unaware of the reaction

> Honestly was not aware of the reaction part... Seeing as a report hides the post immediately, "my" reaction is hidden from me as well.  

https://njump.me/note1xr26qswwh5phss7kvph6crzwd9q283aa5zecnppyy4j82gdszalqf7k67w

![image](https://github.com/user-attachments/assets/8ad4a61f-3d18-4085-8596-8cb21bc01a81)

**Comment by vitorpamplona** [comment]
>> The two things can and should happen in parallel because they are feeding different systems with separate semantics.
>> There is no additional semantics that we build around the warning emoji.
> Which is it? 

What I meant is that the report is doing report things, the reaction is doing reaction things. There is no **additional** semantics being added just because Amethyst does that automatically. The report is supposed to be interpreted as a report, the reaction should be interpreted as a reaction. There is no contradiction.

**Comment by vitorpamplona** [comment]
> there is real damage being done by the warning emoji, because people do view feeds of kind 7s in several different clients, in several different ways. It's also unavoidably an overloading of user-generated content with application semantics.

Sorry, but I think the real damage is calling a reactions tab a "Like" tab. Snort even shows `-` reactions as "Like". Other clients ignore the emoji and just present as like as well, which should never happen. That's their problem, not the problem of who is sending custom reactions. 

Coracle only shows `+` in the Like tab, which is the correct way of doing it if you just want to show actual likes. 

Primal converts 🤙 into ❤️, which are not the same thing, while ignoring all of the other reactions.

noStrudel calls everything, including `-` and 👎, a "Like". 

nostter sends a `+` when you select an emoji. ???

Nostur also shows `-` and any other emoji, including negative ones, as a Like 

These broken implementations are the ones creating this issue because people think a reaction is a like, which never was and will never be. If your client wants to show likes, either filter by `+` or create your system of determining if emojis are positive or negative.

**Comment by staab** [comment]
I've said like 3 times that I agree with that point, but it's orthogonal to this discussion. There are two issues that make the worst case scenario possible: publishing reports as reactions, and reading reactions as likes. That combination results in the Very Bad Things of reports being represented as "likes". Even if we convert all like-based clients to emoji-clients, that would still result in the Pretty Bad Thing of reports being represented as reactions still occurring. On the other hand, not publishing reports as reactions would solve both scenarios.

Also consider that as far as I know, Amethyst is the only one using reactions for reports. You have the ability to completely solve this problem by just not publishing reactions on behalf of your users, but you seem to be advocating for a partial solution instead, which requires 50% of the clients in the space to update how they handle reactions (again, not something I'm opposed to, but it's a different discussion).

**Comment by vitorpamplona** [comment]
>  publishing reports as reactions

Again, the warning emoji is **not a report**. We also send a report. The report is a report. The warning emoji is a **reaction** and should be interpreted as a reaction. Repeating your incorrect conclusions on how you think Amethyst works doesn't help.

> On the other hand, not publishing reports as reactions would solve both scenarios.

No, it doesn't. Broken clients will still show negative reactions as "Likes", which Amethyst ALSO sends a lot of. It's just a matter of time until this discussion restarts again on separate action.

**Comment by staab** [comment]
> The warning emoji is a reaction and should be interpreted as a reaction.

I wouldn't disagree with you if users knew these were being published, and were consenting to it. But as it is, you're basically speaking on behalf of your users, because reactions are a type of human speech.

> Broken clients will still show negative reactions as "Likes"

Negative reactions are very different from reports. Reporting CSAM is very different from reacting negatively to it. If someone wants to react with a 🤮 emoji to a post with CSAM in it, that's a reaction, and should be treated as such. But that is not what I understand your users to be intending.

**Comment by vitorpamplona** [comment]
> Negative reactions are very different from reports. 

I agree. I never said that they are equivalent. I don't even understand why you are making this point. 

We send both so that:
- If your client supports a report, it should use the report and do what you must with the post.
- If it doesn't, your users will still see that lots of people don't like this post.

What you can't do is to assume the warning emoji to be a positive emoji and then ignore the report just because there is a reaction to it. That would be idiotic.

**Comment by alltheseas** [comment]
I created a new nostrability issue for discussion of separating the like counter from counting reactions of any emoji 

https://github.com/nostrability/nostrability/issues/89

**Comment by staab** [comment]
Created a PR to fix this: https://github.com/nostr-protocol/nips/pull/1486

**Comment by alltheseas** [comment]
Continued discussion here

https://njump.me/note10u3h2q5mfw9rh4j95t4tx8fn509xhfkyk6p7u72xkea97xu0ak0sjvqh3g
